### PR TITLE
Add starter reconcile function signatures

### PR DESCRIPTION
**Pull request by leelavg**

### CODE REVIEW

File: pkg/reconciler/reconcile.go [hunk 1]
```diff
@@ -9,6 +9,36 @@ import (
...
 
+type FleetReconciler struct {
+	// Data required for reconcile phases
+	// For AWS (concurrent safe) & IBM Clients decide whether they are created
+	// and destroyed per each reconcile or globals can be used
+}
+
+func (r *FleetReconciler) initReconciler(ctx context.Context) {
+	// Initializes the fields of FleetReconciler struct
+}
```
Comment (nb-ohad) [inline pkg/reconciler/reconcile.go]: No need to add artificial code structures before we observe a use for them.

File: pkg/reconciler/reconcile.go [hunk 2]
```diff
@@ -9,6 +9,36 @@ import (
...
+
+func (r *FleetReconciler) reconcilePhases() (forman.Result, error) {
+	// Reconciles the desired state
+	// Example for an installation flow:
+	// 	1. Ensures the infra via AWS Cloudformation
+	// 	2. Ensures ROKS Cluster installation on existing infra
+	// 	3. Ensures ManagedFusionAgent Operator on ROKS Cluster
+	// 	4. Ensures presence of config for MFAO monitoring stack
+	// 	5. Ensures Data Foundation Config is present on ROKS Cluster
+	return forman.Result{}, nil
```
Comment (nb-ohad) [inline pkg/reconciler/reconcile.go]: As far as I am concerned this should be the content of `r.Reconcile`. Why do we need this second method for?

File: pkg/reconciler/reconcile.go [hunk 3]
```diff
@@ -9,7 +9,35 @@ import (
...
+}
+
+func (r *FleetReconciler) Reconcile(ctx context.Context) (forman.Result, error) {
+	r.initReconciler(ctx)
+	result, err := r.reconcilePhases()
+	if err != nil {
+		return forman.Result{}, err
+	}
+	return result, nil
+}
```
Comment (nb-ohad) [inline pkg/reconciler/reconcile.go]: This entire function is equivalent to:
```
func (r *FleetReconciler) Reconcile(ctx context.Context) (forman.Result, error) {
	r.initReconciler(ctx)
	return r.reconcilePhases()
}
```

As I already mentioned that there is not yet a reason to have `initReconciler` in a prev comment then I would say it is equivalent to 
```
func (r *FleetReconciler) Reconcile(ctx context.Context) (forman.Result, error) {
	return r.reconcilePhases()
}
```

Which means we don't need r.reconcilePhases() and we can just fold this into Reconcile.

**Review by nb-ohad — CHANGES_REQUESTED** [review]

File: pkg/reconciler/reconcile.go (hunk 1 above)
Comment (leelavg) [reply pkg/reconciler/reconcile.go]: done.

File: pkg/reconciler/reconcile.go (hunk 2 above)
Comment (leelavg) [reply pkg/reconciler/reconcile.go]: done.

File: pkg/reconciler/reconcile.go (hunk 3 above)
Comment (leelavg) [reply pkg/reconciler/reconcile.go]: done.

File: pkg/reconciler/reconcile.go [hunk 4]
```diff
@@ -9,7 +9,22 @@ import (
...
+
+func (r *FleetReconciler) Reconcile(ctx context.Context) (forman.Result, error) {
+	// Reconciles the desired state
+	// Example for an installation flow:
+	// 	1. Ensures the infra via AWS Cloudformation
+	// 	2. Ensures ROKS Cluster installation on existing infra
+	// 	3. Ensures ManagedFusionAgent Operator on ROKS Cluster
+	// 	4. Ensures presence of config for MFAO monitoring stack
+	// 	5. Ensures Data Foundation Config is present on ROKS Cluster
+	return forman.Result{}, nil
```
Comment (nb-ohad) [inline pkg/reconciler/reconcile.go:29]: Can we at least save the context on the reconciler?

Comment (leelavg) [reply pkg/reconciler/reconcile.go:29]: - I might need to wrap zap logger with logr and pass that interface for logging, can I go head or do you want simple `fmt` for now?

Comment (nb-ohad) [reply pkg/reconciler/reconcile.go:29]: @Madhu-1 already checked that, is seems that the logr interface is part of the operator framework which I think we should not introduce a dependency one. Can you just pass on the actual zap logger?

Comment (Madhu-1) [reply pkg/reconciler/reconcile.go:29]: here is the PR for it, https://github.com/red-hat-storage/managed-fusion-fleet-reconciler/pull/9, I looked at it with a fresh mind and found a easy solution 🗡️

Comment (nb-ohad) [reply pkg/reconciler/reconcile.go:29]: Great! I have looked at that PR and have a single [request](https://github.com/red-hat-storage/managed-fusion-fleet-reconciler/pull/9#issuecomment-1651422894)

File: main.go [hunk 5]
```diff
@@ -98,9 +98,12 @@ func main() {
 	}
 	defer dbClient.Close(ctx)
 
+	reconcile := &reconciler.FleetReconciler{
+		// data required only once and is concurrent safe
+	}
 	reqChan := forman.GoForman(logger, conf.Reconcile.Concurrency,
 		func(req forman.Request) forman.Result {
-			return reconciler.Reconcile(logger, dbClient, req)
+			return reconciler.Reconcile(logger, reconcile, dbClient, req)
```
Comment (nb-ohad) [inline main.go]: This is wrong, every reconciliation should have its reconciler instance to hold the specific reconcile context.

File: pkg/reconciler/reconcile.go [hunk 6]
```diff
@@ -9,7 +9,22 @@ import (
...
+}
+
+func Reconcile(logger *zap.Logger, r *FleetReconciler, client *db.Database, req forman.Request) forman.Result {
```
Comment (nb-ohad) [inline pkg/reconciler/reconcile.go]: Getting the reconciler from the outside does not make sense, as it should be a local instance to this specific reconcile to hold this specific reconcile context

File: pkg/reconciler/reconcile.go [hunk 7]
```diff
@@ -19,5 +34,12 @@ func Reconcile(logger *zap.Logger, client *db.Database, req forman.Request) form
 	}
 	logger.Info("Provider", zap.String("clusterId", provider.ClusterID))
 
-	return forman.Result{}
+	result, err := r.Reconcile(ctx)
```
Comment (nb-ohad) [inline pkg/reconciler/reconcile.go]:
```suggestion
	r := FleetReconciler{
		ctx: ctx,
	}
	result, err := r.Reconcile()
```

**Review by nb-ohad — CHANGES_REQUESTED** [review]

File: main.go (hunk 5 above)
Comment (leelavg) [reply main.go]: - There are some resources which are better instantiated once and re-used for every reconcile
- At the moment, those are AWS & IBM Client.

File: pkg/reconciler/reconcile.go (hunk 7 above)
Comment (leelavg) [reply pkg/reconciler/reconcile.go]: Ack, however we are attaching the supplied ctx at the start of Reconcile method.

File: main.go (hunk 5 above)
Comment (leelavg) [reply main.go]: - On another thought, I think I'm wrong due to not taking concurrency into account and sharing same struct with multiple workers🤦
- To be clean I don't see a way (yet) on using fields which can be instantiated once 🤔

Comment (nb-ohad) [reply main.go]: We already have that, good examples are logger and dbClient. They are instantiated once and used in every reconciliation invocation. The critical thing is that they need to be either stateless, immutable, or concurrency-safe.  
The point for the reconciler instance is to be a local context for a specific reconcile

Comment (leelavg) [reply main.go]: - yes, the FleetReconciler doesn't tick the requirements to be instantiated at once and so need to supply those via Reconcile function.

Comment (leelavg) [reply main.go]: done.

File: pkg/reconciler/reconcile.go (hunk 6 above)
Comment (leelavg) [reply pkg/reconciler/reconcile.go]: done.

File: pkg/reconciler/reconcile.go (hunk 7 above)
Comment (leelavg) [reply pkg/reconciler/reconcile.go]: done.

File: pkg/reconciler/reconcile.go [hunk 8]
```diff
@@ -9,6 +9,22 @@ import (
 	"github.com/go-logr/logr"
 )
 
+type FleetReconciler struct {
+	ctx context.Context
+	// Data required for reconcile phases
+}
+
+func (r *FleetReconciler) Reconcile() (forman.Result, error) {
```
Comment (Madhu-1) [inline pkg/reconciler/reconcile.go]: struct and method need to be exposed? most of the time we do name the receiver as the first letter of the struct if it makes sense to change it to `f`

File: pkg/reconciler/reconcile.go [hunk 9]
```diff
@@ -9,6 +9,22 @@ import (
 	"github.com/go-logr/logr"
 )
 
+type FleetReconciler struct {
+	ctx context.Context
+	// Data required for reconcile phases
```
Comment (Madhu-1) [inline pkg/reconciler/reconcile.go]: Dont we need logger or dbClient as members or input to Reconcile function?

File: pkg/reconciler/reconcile.go (hunk 8 above)
Comment (leelavg) [reply pkg/reconciler/reconcile.go]: done.

File: pkg/reconciler/reconcile.go (hunk 9 above)
Comment (leelavg) [reply pkg/reconciler/reconcile.go]: for now logger is fine, I don't want db client to be part of this PR, done.

File: pkg/reconciler/reconcile.go [hunk 10]
```diff
@@ -19,5 +38,17 @@ func Reconcile(log logr.Logger, client *db.Database, req forman.Request) forman.
 	}
 	log.Info("Provider", "clusterId", provider.ClusterID)
 
-	return forman.Result{}
+	r := &fleetReconciler{
+		ctx:          ctx,
+		log:          log,
+		providerName: provider.ClusterID,
```
Comment (nb-ohad) [inline pkg/reconciler/reconcile.go]: The provider record is already loaded (from the DB) so we should save it on the reconciler not just the name 
```suggestion
		provider: provider,
```

**Review by nb-ohad — CHANGES_REQUESTED** [review]

File: pkg/reconciler/reconcile.go (hunk 10 above)
Comment (leelavg) [reply pkg/reconciler/reconcile.go]: done.

**Review by nb-ohad — APPROVED** [review]

**Pull request merged**